File: simulation.py

from collections import Counter
import sys

import numpy as np

from intersection import Intersection
from street import Street
from car import Car

from stable_priority_queue import StablePriorityQueue

class Simulation:
    def __init__(
        self, 
        duration: int, 
        intersections: list[Intersection], 
        streets: list[Street], 
        cars: list[Car], 
        street_mapping: dict[str, Street], 
        bonus: int
        ):
        self.duration = duration
        self.intersections = intersections
        self.streets = streets
        self.street_mapping = street_mapping
        self.cars = cars
        self.bonus = bonus

    @classmethod
    def from_file(cls, filename):
        if filename is not None:
            with open(filename, 'r') as file:
                return cls._from_file(file)
        else:
            return cls._from_file(sys.stdin)

    @classmethod
    def _from_file(cls, file):
        D, I, S, V, F = file.readline().split()
        duration = int(D)
        num_of_intersections = int(I)
        num_of_streets = int(S)
        num_of_cars = int(V)
        bonus = int(F)
        intersections = np.empty(num_of_intersections, dtype=Intersection)
        streets = np.empty(num_of_streets, dtype=Street)
        street_mapping = {}
        for i in range(num_of_streets):
            start, end, name, length = file.readline().split()
            start = int(start)
            end = int(end)
            length = int(length)
            if intersections[start] is None:
                intersections[start] = Intersection(start)
            if intersections[end] is None:
                intersections[end] = Intersection(end)
            street = Street(i, intersections[start], intersections[end], name, length)
            streets[i] = street
            street_mapping[street.name] = street
            street.start.outgoing.append(street)
            street.end.incoming.append(street)

        cars = np.empty(num_of_cars, dtype=Car)
        for i in range(num_of_cars):
            street_names = file.readline().split()[1:]
            path = [street_mapping[name] for name in street_names]
            car = Car(i, len(path), path)
            cars[i] = car
            for street in car.path:
                street.used = True

        return cls(duration, intersections, streets, cars, street_mapping, bonus)

    def __str__(self):
        lines = []
        lines.append(f'Duration: {self.duration}')
        lines.append(f'Number of intersections: {len(self.intersections)}')
        lines.append(f'Bonus points: {self.bonus}')
        lines.append(f'Streets: {len(self.streets)}')
        lines.extend(map(str, self.streets))
        lines.append(f'Cars: {len(self.cars)}')
        lines.extend(map(str, self.cars))
        return '\n'.join(lines)

    def run(self):
        time = 0
        event_queue = StablePriorityQueue(maxsize=1000)
        for car in self.cars:
            street = car.current_street()
            street.queueing_cars.put(car)
            if street.queueing_cars.qsize() == 1:
                event_queue.put((street.next_green(time), street))

        while not event_queue.empty():
            time, object = event_queue.get()
            if time > self.duration:
                break
            if isinstance(object, Street):
                street = object
                car = street.go(time)
                event_queue.put((time + car.current_street().length, car))
                if not street.queueing_cars.empty():
                    event_queue.put((street.next_green(time), street))

            elif isinstance(object, Car):
                car = object
                if car.is_final_destination():
                    car.finished = True
                    car.finish_time = time
                else:
                    street = car.current_street()
                    street.queueing_cars.put(car)
                    if street.queueing_cars.qsize() == 1:
                        event_queue.put((street.next_green(time), street))              

        return self

    def score(self, verbose=False):
        if verbose:
            return self._score_verbose()
        
        scores = np.zeros(len(self.cars), dtype=int)
        for i, car in enumerate(self.cars):
            if car.finished:
                scores[i] = self.bonus + (self.duration - car.finish_time)
        return np.sum(scores)

    def _score_verbose(self):
        scores = np.zeros(len(self.cars), dtype=int)
        finish_times = np.zeros(len(self.cars), dtype=int)
        for i, car in enumerate(self.cars):
            if car.finished:
                scores[i] = self.bonus + (self.duration - car.finish_time)
                finish_times[i] = car.finish_time
        avg_ride_time = np.mean(finish_times[finish_times > 0])
        scores_masked = np.ma.MaskedArray(scores, scores == 0)
        finished = scores_masked.count()
        max_idx = np.argmax(scores_masked)
        min_idx = np.argmin(scores_masked)
        score = np.sum(scores)
        print('#' * 80)
        print(f'The submission scored **{score:,} points**.') 
        print(f'This is the sum of {finished * self.bonus:,} bonus points for ' \
            f'cars arriving before the deadline ({self.bonus:,} points each) and '\
            f'{score - finished * self.bonus:,} points for early arrival times.')
        print(f'{finished} of {len(self.cars)} cars arrived before the deadline ' \
            f'({int(finished / len(self.cars) * 100):,} %).')
        print(f'The earliest car (ID {max_idx}) arrived at its destination after ' \
            f'{finish_times[max_idx]} seconds scoring {scores[max_idx]:,} points,')
        print(f'whereas the last car (ID {min_idx}) arrived at its destination ' \
            f'after {finish_times[min_idx]} seconds scoring {scores[min_idx]:,} points.')
        print(f'Cars that arrived within the deadline drove for an average of ' \
            f'{avg_ride_time:0.2f} seconds to arrive at their destination.')
        return score

    def read_plan(self, filename):
        with open(filename, 'r') as file:
            num_of_intersections = int(file.readline())
            for _ in range(num_of_intersections):
                intersection_id = int(file.readline())
                num_of_streets = int(file.readline())

                intersection = self.intersections[intersection_id]
                plan = intersection.plan
                for _ in range(num_of_streets):
                    street_name, green_time = file.readline().split()
                    green_time = int(green_time)

                    street = self.get_street_by_name(street_name)
                    street._green_interval = range(len(plan), len(plan) + green_time)
                    plan.extend([street] * green_time)
                    
        return self

    def create_plan_default(self):
        """
        For every intersection and every incoming street, set green for 1 second.
        """
        for intersection in self.intersections:
            plan = intersection.plan
            for street in intersection.incoming:
                street._green_interval = range(len(plan), len(plan) + 1)
                plan.append(street)

        return self

    def create_plan_used(self):
        """
        Set green for 1 second for every used incoming street. 
        Exclude streets that aren't used.
        """
        # TODO: Maybe not ideal, try to fix this
        for intersection in self.intersections:
            for street in intersection.incoming:
                if street.used:
                    plan = intersection.plan
                    street._green_interval = range(len(plan), len(plan) + 1)
                    plan.append(street)

        return self

    def create_output(self, filename):
        if filename is not None:
            with open(filename, 'w') as file:
                self._create_output(file)
        else:
            self._create_output(sys.stdout)

    def _create_output(self, file):
        # TODO: Not ideal, try to come up with a better solution
        intersections_used = []
        for intersection in self.intersections:
            if intersection.plan:
                intersections_used.append(intersection)

        file.write(str(len(intersections_used)) + '\n')

        for intersection in intersections_used:
            file.write(str(intersection.id) + '\n')
            counter = Counter(intersection.plan)
            file.write(str(len(counter)) + '\n')
            for street, time in counter.items():
                file.write(f'{street.name} {str(time)}\n')

    def get_street_by_name(self, name):
        return self.street_mapping[name]

File: simulation/src/simulation.cpp
#include <algorithm>
#include <fstream>
#include <functional>
#include <iomanip>
#include <ios>
#include <iostream>
#include <limits>
#include <memory>
#include <numeric>
#include <queue>
#include <string>
#include <vector>

#include "simulation.hpp"

Simulation::Simulation(const std::string &filename) {
    std::ifstream data_file{filename};
    size_t duration;
    size_t number_of_intersections;
    size_t number_of_streets;
    size_t number_of_cars;
    size_t bonus;

    data_file >> duration >> number_of_intersections >> number_of_streets >> number_of_cars >> bonus;

    duration_ = duration;
    bonus_ = bonus;

#ifdef DEBUG
    std::cout << duration_ << " " << number_of_intersections << " "
              << number_of_streets << " " << number_of_cars << " "
              << bonus_ << "\n";
#endif

    intersections_.reserve(number_of_intersections);
    for (size_t id = 0; id < number_of_intersections; ++id) {
        intersections_.emplace_back(id);
    }

    streets_.reserve(number_of_streets);
    cars_.reserve(number_of_cars);

    size_t start;
    size_t end;
    std::string name;
    size_t length;
    for (size_t id = 0; id < number_of_streets; ++id) {
        data_file >> start >> end >> name >> length;
#ifdef DEBUG
        std::cout << start << " " << end << " " << name << " " << length << "\n";
#endif
        auto &&street = streets_.emplace_back(id, start, end, name, length);
        street_mapping_.emplace(street.name(), id);
        intersections_[end].add_incoming(id);
    }


    size_t path_length;
    std::string street_name;

    for (size_t id = 0; id < number_of_cars; ++id) {
        data_file >> path_length;
#ifdef DEBUG
        std::cout << path_length << " ";
#endif
        std::vector<size_t> path;
        for (size_t i = 0; i < path_length; ++i) {
            data_file >> street_name;
#ifdef DEBUG
            std::cout << street_name << " ";
#endif
            auto &&street_id = street_mapping_[street_name];
            path.emplace_back(street_id);

            // The last street in path is not marked as used because it doesn't
            // use the traffic light
            if (i < path_length - 1) {
                streets_[street_id].set_used(true);
            }
        }
        cars_.emplace_back(id, path_length, path);
#ifdef DEBUG
        std::cout << "\n";
#endif
    }
}

const std::vector<Intersection> &Simulation::intersections() const {
    return intersections_;
}

const std::vector<Street> &Simulation::streets() const {
    return streets_;
}

const std::vector<Car> &Simulation::cars() const {
    return cars_;
}

std::ostream &operator<<(std::ostream &os, const Simulation &obj) {
    os << "Duration: " << obj.duration_ << "\n"
       << "Number of intersections: " << obj.intersections_.size() << "\n"
       << "Bonus: " << obj.bonus_ << "\n"
       << "Number of streets: " << obj.streets_.size() << "\n";
    for (auto &&s: obj.streets_) {
        os << s << "\n";
    }
    os << "Number of cars: " << obj.cars_.size() << "\n";
    for (auto &&c: obj.cars_) {
        os << c << "\n";
    }
    return os;
}

size_t Simulation::duration() const {
    return duration_;
}

size_t Simulation::bonus() const {
    return bonus_;
}

Simulation::Instance Simulation::create_instance() {
    return Simulation::Instance{*this};
}

Simulation::Instance::Instance(const Simulation &data) : data_(data) {
    intersections_.reserve(data.intersections().size());
    streets_.reserve(data.streets().size());
    cars_.reserve(data.cars().size());

    for (auto &&i: data.intersections()) {
        intersections_.emplace_back(i);
    }
    for (auto &&s: data.streets()) {
        streets_.emplace_back(s);
    }
    for (auto &&c: data.cars()) {
        cars_.emplace_back(c);
    }
}

void Simulation::Instance::read_plan(const std::string &filename) {
    std::ifstream file{filename};
    size_t number_of_intersections;
    size_t intersection_id;
    size_t number_of_streets;
    std::string street_name;
    size_t green_light_duration;

    file >> number_of_intersections;
    for (size_t i = 0; i < number_of_intersections; ++i) {
        file >> intersection_id;
        schedules_.emplace(intersection_id, intersection_id);

        file >> number_of_streets;
        for (size_t j = 0; j < number_of_streets; ++j) {
            file >> street_name >> green_light_duration;
            auto &&street_id = data_.street_mapping_.at(street_name);
            schedules_.at(intersection_id).add_street(street_id, green_light_duration);
        }
    }
}

void Simulation::Instance::write_plan(const std::string &filename) {
    std::ofstream file{filename};
    file << schedules_.size() << "\n";

    for (auto &&[intersection_id, schedule]: schedules_) {
        file << intersection_id << "\n"
             << schedule.length() << "\n";
        //TODO: finish this
    }
    //for (const Intersection::Instance &s: scheduled) {
    //    auto &&schedule = s.schedule().value().get();
    //    file << s.id() << "\n"
    //         << schedule.length() << "\n";
    //
    //    std::vector<std::pair<size_t, std::ranges::iota_view<size_t, size_t>>> schedule_map{schedule.cbegin(), schedule.cend()};
    //    auto cmp = [](const auto &lhs, const auto &rhs) {
    //        return *lhs.second.begin() < *rhs.second.begin();
    //    };
    //    std::sort(schedule_map.begin(), schedule_map.end(), cmp);
    //
    //    for (auto &&[street_id, range]: schedule_map) {
    //        auto &&street_name = streets_[street_id].name();
    //        auto green_light_duration = (*range.end() - *range.begin());
    //        file << street_name << " " << green_light_duration << "\n";
    //    }
    //}
}

void Simulation::Instance::create_plan_default() {
    for (auto &&intersection: intersections_) {
        schedules_.emplace(intersection.id(), intersection.id());
        for (auto &&s: intersection.incoming()) {
            auto &&street = streets_[s];
            if (street.is_used()) {
                schedules_.at(intersection.id()).add_street(street.id(), 1);
            }
        }
    }
}

void Simulation::Instance::initialize_event_queue(auto &event_queue) {
    for (auto &&car: cars_) {
        auto &&street = streets_[car.current_street()];
        street.add_car(car.id());
        if (street.car_queue_size() == 1) {

            auto &&intersection = intersections_[street.end()];
            auto &&next_green_time = schedules_.at(intersection.id()).next_green(street, 0);
            if (next_green_time.has_value()) {
                event_queue.emplace(std::make_unique<StreetEvent>(next_green_time.value(), street));
            }
        }
    }
}

void Simulation::Instance::process_street_event(auto &event_queue, auto &event) {
    auto &&street_event = static_cast<StreetEvent *>(event.get());
    auto time = street_event->time();
    auto &&street = street_event->street();
    auto &&car = cars_[street.get_car(time)];
    car.move_to_next_street();

    event_queue.pop();
    event_queue.emplace(std::make_unique<CarEvent>(time + streets_[car.current_street()].length(), car));

    if (street.car_queue_size() > 0) {

        auto &&intersection = intersections_[street.end()];
        auto &&next_green_time = schedules_.at(intersection.id()).next_green(street, time);
        if (next_green_time.has_value()) {
            event_queue.emplace(std::make_unique<StreetEvent>(next_green_time.value(), street));
        }
    }
}

void Simulation::Instance::process_car_event(auto &event_queue, auto &event) {
    auto &&car_event = static_cast<CarEvent *>(event.get());
    auto time = car_event->time();
    auto &&car = static_cast<CarEvent *>(event.get())->car();
    event_queue.pop();
    if (car.at_final_destination()) {
        car.set_finished(true);
        car.set_finish_time(time);
    } else {
        auto &&street = streets_[car.current_street()];
        street.add_car(car.id());
        if (street.car_queue_size() == 1) {

            auto &&intersection = intersections_[street.end()];
            auto &&next_green_time = schedules_.at(intersection.id()).next_green(street, time);
            if (next_green_time.has_value()) {
                event_queue.emplace(std::make_unique<StreetEvent>(next_green_time.value(), street));
            }
        }
    }
}

Simulation::Instance &Simulation::Instance::run() {
    // an earlier event has higher priority
    auto cmp = [](const std::unique_ptr<Event> &lhs, const std::unique_ptr<Event> &rhs) {
        return !(*lhs < *rhs);
    };
    std::priority_queue<std::unique_ptr<Event>, std::vector<std::unique_ptr<Event>>, decltype(cmp)> event_queue;
    //TODO: finish refactoring the methods below
    initialize_event_queue(event_queue);

    while (!event_queue.empty()) {
        auto &&event = event_queue.top();
        if (event->time() > data_.duration_) {
            break;
        }

        if (event->event_type() == Event::STREET_EVENT_TYPE) {
            process_street_event(event_queue, event);
        } else if (event->event_type() == Event::CAR_EVENT_TYPE) {
            process_car_event(event_queue, event);
        }
    }
    return *this;
}

size_t Simulation::Instance::score(bool verbose) const {
    if (!verbose) {
        auto count_score = [this](size_t total, const Car::Instance &car) {
            if (car.finished()) {
                total += this->data_.bonus_ + this->data_.duration_ - car.finish_time();
            }
            return total;
        };
        return std::accumulate(cars_.cbegin(), cars_.cend(), 0, count_score);
    }
    size_t finished_count = 0;
    size_t total_ride_time = 0;
    size_t total_score = 0;
    size_t max_score = 0;
    size_t min_score = std::numeric_limits<size_t>::max();
    const Car::Instance *earliest_car = nullptr;
    const Car::Instance *latest_car = nullptr;
    for (auto &&c: cars_) {
        if (c.finished()) {
            ++finished_count;
            total_ride_time += c.finish_time();
            auto score = data_.bonus_ + data_.duration_ - c.finish_time();
            total_score += score;
            if (score > max_score) {
                max_score = score;
                earliest_car = &c;
            }
            if (score < min_score) {
                min_score = score;
                latest_car = &c;
            }
        }
    }
    float average_ride_time = static_cast<float>(total_ride_time) / static_cast<float>(finished_count);

    // return int as a thousand seperated string
    auto thousand_sep = [](size_t i) {
        std::string s = std::to_string(i);
        std::string formatted;

        int count = 0;
        for (int pos = static_cast<int>(s.length()) - 1; pos >= 0; --pos) {
            formatted.insert(0, 1, s[static_cast<size_t>(pos)]);
            ++count;
            if (count % 3 == 0 && pos > 0) {
                formatted.insert(0, 1, ',');
            }
        }
        return formatted;
    };

    std::cout << "The submission scored **" << thousand_sep(total_score) << " points**. "
              << "This is the sum of " << thousand_sep(finished_count * data_.bonus_) << " bonus points for "
              << "cars arriving before the deadline (" << thousand_sep(data_.bonus_) << " points each) and "
              << thousand_sep(total_score - finished_count * data_.bonus_) << " points for early arrival times.\n\n"
              << thousand_sep(finished_count) << " of " << thousand_sep(cars_.size()) << " cars arrived before the deadline "
              << "(" << static_cast<float>(finished_count) / static_cast<float>(cars_.size()) * 100 << "%). "
              << "The earliest car (ID " << earliest_car->id() << ") arrived at its destination after "
              << thousand_sep(earliest_car->finish_time()) << " seconds scoring " << thousand_sep(max_score) << " points, "
              << "whereas the last car (ID " << latest_car->id() << ") arrived at its destination "
              << "after " << thousand_sep(latest_car->finish_time()) << " seconds scoring " << thousand_sep(min_score) << " points. "
              << "Cars that arrived within the deadline drove for an average of " << std::fixed
              << std::setprecision(2) << average_ride_time << " seconds to arrive at their destination.\n";
    return total_score;
}

const std::vector<Intersection::Instance> &Simulation::Instance::intersections() const {
    return intersections_;
}

const std::vector<Street::Instance> &Simulation::Instance::streets() const {
    return streets_;
}

const std::vector<Car::Instance> &Simulation::Instance::cars() const {
    return cars_;
}
